### DISCUSSION TITLE
Detecting screen/text/button during connection

### DISCUSSION THREAD

**Discussion by jayprojects** [FreeRDP/FreeRDP#10505] [General]
I'm trying to detect "Windows Legal Notice banner" screen during/after login.

I checked wfContext and others for clue (in wf_client.c), like `wfc->window_title`, `wfUiInfo.pscCaptionText`, `wfUiInfo.pszMessageText `they are always static regardless of progress/change on rdp screen.

I can get bitmap from `wfc->drawing` or `wfc->primary` in wf_client_thread method or `gdi->primary_buffer` in wf_end_paint method. and i can see the Legal Notice screen on the saved bitmap. but Image processing a raw bitmap for screen/text detection can be resource intensive. 

Is there any other way to get a sense of the window/content of the rdp screen?

Thanks for your help.

**Comment by akallabeth** [comment]
@jayprojects
* what is this `Windows Legal Notice Banner` you mention? (too ambiguous, could be many things)
* as a rule of thumb, as long as it is not something separate from the normal `RDP` stream image recognition is the only way to go

**Reply by jayprojects** [reply]
@akallabeth
Thank you for your response. Windows Legal Notice Banner is an acknowledge window/screen. with an "OK" button and some text. It can be enabled or disabled by a registry setting. if enabled, that text with "ok" button shows up right after someone logs in (RDP/ enters valid credential). If the user clicks OK button in some time, it proceeds as usual, if not, (RDP) session gets disconnected.

Here is a sample banner from net
![screenshot-2022-11-12-112453](https://github.com/user-attachments/assets/08f28ae8-f11f-441a-854f-0a853e5f45c6)

Yes as i mention, image processing would be the last option, Just checking if FreerRDP has some textual information on the current window that is being displayed. either title or code or id.

OR if FreerRDP has a smaller version of the screen image data. like a thumbnail? the` wfc->drawing, wfc->primary, gdi->primary_buffer` they contain bitmap of the full screen in high resolution. Data array is large. It would be resource intensive to work with that. I was hoping for smaller/lighter image.

Also does anyone know the correct bitmap profile of the image data? i tried following 
```

BITMAPINFOHEADER BMIH; 
BMIH.biSize = sizeof(BITMAPINFOHEADER); 
BMIH.biWidth = w; 
BMIH.biHeight = -1*h; 
BMIH.biPlanes = 1; 
BMIH.biSizelmage = w * h * 3; 
BMIH.biBitCount = 24; 
BMIH.biCompression = BI_RGB; 

BITMAPFILEHEADER bmfh; 
int nBitsOffset = sizeof(BITMAPFILEHEADER); 
LONG llmageSize = BMIH.biSizelmage; 
LONG 1FileSize = nBitsOffset + llmageSize; 
bmfh.bfType = 'B' + ('M' << 8); 
bmfh.bfOffBits = nBitsOffset; 
bmfh.bfSize = 1FileSize; 

UINT nWrittenFileHeaderSize = fwrite(&bmfh, 1, sizeof(BITMAPFILEHEADER), pFile); 
UINT nWrittenlnfoHeaderSize = fwrite(&BMIH, 1, sizeof(BITMAPINFOHEADER), pFile); 
UINT nWrittenDlBDataSize = fwrite(lpBits, 1, llmageSize, pFile); 


```

but resultant bitmap file has very wrong color and text distorted. with `biBitCount = 16` text visible but color is still off. with `biBitCount = 8` color look right but scale is wrong. only get partial screen. 

Thanks again

**Reply by akallabeth** [reply]
@jayprojects 
* Ok, so some nagging dialog within the session. Yes, correct, only image recognition possible.
* No, `FreeRDP` only contains the decoded buffers to display on your local screen in `gdi->primary_buffer`, format is some `PIXEL_FORMAT_*` that you pass to `gdi_init`, you also need `gdi->stride` as the bytesize of a line in that buffer (There are some helper functions to write bitmaps/webp/png/jpeg to file that might help you there. I also assume you are using `wfreerdp` here, as your code looks like it. The buffers might be empty or non existant for clients that do not rely on our software decoding stack or offload `GFX` related stuff)
